File: copilot-api/app/report_suggestions/report_filter_roll_dates.py
"""
Align with copilot-widget-v3/src/app/report-filter-roll-dates.ts:
for Smart-default re-runs, roll end/snapshot date fields to today (MM/DD/YYYY)
while leaving start/from dates unchanged.

- Strict keys (*ToDate, todate, …): any non-empty value is replaced with today (ERP parity).
- Context keys (asof, cutoff, through, …): only values that look like calendar dates are updated.
"""

from __future__ import annotations

import json
import re
from datetime import date

_US_DATE_RE = re.compile(r"^\d{1,2}/\d{1,2}/\d{4}$")
_ISO_DATE_PREFIX_RE = re.compile(r"^\d{4}-\d{2}-\d{2}")

_END_DATE_KEYS: frozenset[str] = frozenset(
    {
        "mailingContactToDate",
        "mailingSaleToDate",
        "todate",
        "toDate",
        "dateTo",
        "DateTo",
        "saleToDate",
        "contactToDate",
    }
)

# Exact keys (lowercase) treated as “as of / snapshot / period end” — only if value looks like a date.
_CONTEXT_DATE_KEYS: frozenset[str] = frozenset(
    {
        "asof",
        "asofdate",
        "as_of",
        "as_of_date",
        "reportasof",
        "report_asof",
        "snapshotdate",
        "snapshot_date",
        "cutoffdate",
        "cut_off_date",
        "cutoff_date",
        "thrudate",
        "throughdate",
        "through_date",
        "periodenddate",
        "period_end_date",
        "endingdate",
        "enddate",
        "end_date",
        "datethrough",
        "date_through",
    }
)


def _looks_like_date_value(s: str) -> bool:
    if not s or not s.strip():
        return False
    t = s.strip()
    if _US_DATE_RE.fullmatch(t):
        return True
    if _ISO_DATE_PREFIX_RE.match(t):
        return True
    return False


def _is_exclude_key(key: str) -> bool:
    """Start / from bounds — never roll to today."""
    k = key.strip()
    if re.search(r"FromDate$", k, re.IGNORECASE):
        return True
    kl = k.lower()
    if kl in (
        "fromdate",
        "from_date",
        "startdate",
        "start_date",
        "begindate",
        "begin_date",
        "datefrom",
        "date_from",
        "mailingcontactfromdate",
    ):
        return True
    if re.search(r"fromdate$", k, re.IGNORECASE):
        return True
    if re.search(r"from_date$", k, re.IGNORECASE):
        return True
    if re.search(r"startdate$", k, re.IGNORECASE):
        return True
    if re.search(r"begindate$", k, re.IGNORECASE):
        return True
    if re.search(r"datefrom$", k, re.IGNORECASE):
        return True
    return False


def _is_end_date_key(key: str) -> bool:
    if key in _END_DATE_KEYS:
        return True
    k = key.strip()
    if re.search(r"todate$", k, re.IGNORECASE):
        return True
    if re.search(r"ToDate$", k, re.IGNORECASE) and not re.search(r"FromDate$", k, re.IGNORECASE):
        return True
    if re.search(r"DateTo$", k, re.IGNORECASE):
        return True
    return False


def _is_context_roll_key(key: str) -> bool:
    """As-of / snapshot / period-end style keys (not strict *ToDate)."""
    if _is_exclude_key(key):
        return False
    k = key.strip()
    kl = k.lower()
    if kl in _CONTEXT_DATE_KEYS:
        return True
    if re.fullmatch(r"asof(_date)?", kl):
        return True
    if re.search(r"(^|_)asof$", kl):
        return True
    if re.search(r"(^|_)asofdate$", kl):
        return True
    if re.search(r"(through|thru)date$", kl):
        return True
    if re.search(r"(cutoff|snapshot)date$", kl):
        return True
    if re.search(r"periodenddate$", kl):
        return True
    if re.search(r"(^|_)endingdate$", kl):
        return True
    return False


def _today_us() -> str:
    d = date.today()
    return f"{d.month:02d}/{d.day:02d}/{d.year:04d}"


def apply_rolling_end_dates_to_filter_data(filter_data: str) -> str:
    """
    Returns updated filter_data JSON string, or original if parse fails.
    """
    if not filter_data or not filter_data.strip():
        return filter_data
    try:
        obj = json.loads(filter_data)
    except json.JSONDecodeError:
        return filter_data
    if not isinstance(obj, dict):
        return filter_data

    today = _today_us()
    changed = False
    for key in list(obj.keys()):
        if _is_exclude_key(key):
            continue
        v = obj[key]
        if v is None:
            continue
        s = str(v).strip()
        if not s:
            continue

        if _is_end_date_key(key):
            # Same as legacy widget: any non-empty end-date value becomes today.
            obj[key] = today
            changed = True
            continue
        if _is_context_roll_key(key):
            if _looks_like_date_value(s):
                obj[key] = today
                changed = True
            continue

    if not changed:
        return filter_data
    try:
        return json.dumps(obj, separators=(",", ":"))
    except (TypeError, ValueError):
        return filter_data

File: copilot-widget-v3/src/app/report-filter-roll-dates.ts
/**
 * For Smart-default re-runs: roll end/snapshot date fields to today (MM/DD/YYYY)
 * while leaving start/from dates unchanged. Kept in sync with
 * copilot-api/app/report_suggestions/report_filter_roll_dates.py.
 */

const US_DATE_RE = /^(\d{1,2})\/(\d{1,2})\/(\d{4})$/;

/** Keys whose values are strict end dates (case-sensitive keys as in ERP JSON). */
const END_DATE_KEYS: ReadonlySet<string> = new Set([
  'mailingContactToDate',
  'mailingSaleToDate',
  'todate',
  'toDate',
  'dateTo',
  'DateTo',
  'saleToDate',
  'contactToDate'
]);

/** Lowercase exact keys: as-of / snapshot / period-end — only if value looks like a date. */
const CONTEXT_DATE_KEYS: ReadonlySet<string> = new Set([
  'asof',
  'asofdate',
  'as_of',
  'as_of_date',
  'reportasof',
  'report_asof',
  'snapshotdate',
  'snapshot_date',
  'cutoffdate',
  'cut_off_date',
  'cutoff_date',
  'thrudate',
  'throughdate',
  'through_date',
  'periodenddate',
  'period_end_date',
  'endingdate',
  'enddate',
  'end_date',
  'datethrough',
  'date_through'
]);

function looksLikeDateValue(s: string): boolean {
  const t = s.trim();
  if (!t) {
    return false;
  }
  if (US_DATE_RE.test(t)) {
    return true;
  }
  if (/^\d{4}-\d{2}-\d{2}/.test(t)) {
    return true;
  }
  return false;
}

function isExcludeKey(key: string): boolean {
  if (/FromDate$/i.test(key)) {
    return true;
  }
  const kl = key.trim().toLowerCase();
  if (
    [
      'fromdate',
      'from_date',
      'startdate',
      'start_date',
      'begindate',
      'begin_date',
      'datefrom',
      'date_from',
      'mailingcontactfromdate'
    ].includes(kl)
  ) {
    return true;
  }
  if (/fromdate$/i.test(key)) {
    return true;
  }
  if (/from_date$/i.test(key)) {
    return true;
  }
  if (/startdate$/i.test(key)) {
    return true;
  }
  if (/begindate$/i.test(key)) {
    return true;
  }
  if (/datefrom$/i.test(key)) {
    return true;
  }
  return false;
}

function isEndDateKey(key: string): boolean {
  if (END_DATE_KEYS.has(key)) {
    return true;
  }
  const k = key.trim();
  if (/todate$/i.test(k)) {
    return true;
  }
  if (/ToDate$/i.test(k) && !/FromDate$/i.test(k)) {
    return true;
  }
  if (/DateTo$/i.test(k)) {
    return true;
  }
  return false;
}

function isContextRollKey(key: string): boolean {
  if (isExcludeKey(key)) {
    return false;
  }
  const kl = key.trim().toLowerCase();
  if (CONTEXT_DATE_KEYS.has(kl)) {
    return true;
  }
  if (/^asof(_date)?$/i.test(kl)) {
    return true;
  }
  if (/(^|_)asof$/i.test(kl)) {
    return true;
  }
  if (/(^|_)asofdate$/i.test(kl)) {
    return true;
  }
  if (/(through|thru)date$/i.test(kl)) {
    return true;
  }
  if (/(cutoff|snapshot)date$/i.test(kl)) {
    return true;
  }
  if (/periodenddate$/i.test(kl)) {
    return true;
  }
  if (/(^|_)endingdate$/i.test(kl)) {
    return true;
  }
  return false;
}

function todayUs(): string {
  const d = new Date();
  const mm = String(d.getMonth() + 1).padStart(2, '0');
  const dd = String(d.getDate()).padStart(2, '0');
  const yyyy = d.getFullYear();
  return `${mm}/${dd}/${yyyy}`;
}

/**
 * Returns updated filter_data JSON string, or original if parse fails.
 */
export function applyRollingEndDatesToFilterData(filterData: string | null | undefined): string {
  if (!filterData || !filterData.trim()) {
    return filterData || '';
  }
  let obj: Record<string, unknown>;
  try {
    obj = JSON.parse(filterData) as Record<string, unknown>;
  } catch {
    return filterData;
  }
  if (!obj || typeof obj !== 'object' || Array.isArray(obj)) {
    return filterData;
  }
  const today = todayUs();
  let changed = false;
  for (const key of Object.keys(obj)) {
    if (isExcludeKey(key)) {
      continue;
    }
    const v = obj[key];
    if (v === null || v === undefined) {
      continue;
    }
    const s = String(v).trim();
    if (!s) {
      continue;
    }
    if (isEndDateKey(key)) {
      if (US_DATE_RE.test(s)) {
        obj[key] = today;
        changed = true;
        continue;
      }
      if (/^\d{4}-\d{2}-\d{2}/.test(s)) {
        obj[key] = today;
        changed = true;
        continue;
      }
      obj[key] = today;
      changed = true;
      continue;
    }
    if (isContextRollKey(key)) {
      if (looksLikeDateValue(s)) {
        obj[key] = today;
        changed = true;
      }
      continue;
    }
  }
  if (!changed) {
    return filterData;
  }
  try {
    return JSON.stringify(obj);
  } catch {
    return filterData;
  }
}
